Kconfig: CMakeLists: Define building of board images

rgundi opened the pull request:
This patchset introduces support for building board specific images.
This may differ from board to board but may generally involve padding
the output binary with some additional data.

This patchset illustrates that for Intel S1000 board.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
intel_s1000_crb: CMakeLists: Define building of custom image

This patchset introduces support for building a board specific custom
image for intel_s1000_crb. This pads the output binary (zephyr.bin)
with some additional data as required by the s1000 BootROM.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

nashif commented on Kconfig.zephyr:
```diff
@@ -306,6 +306,13 @@ config BUILD_OUTPUT_STRIPPED
 	  Build a stripped binary. This will build a zephyr.stripped file need
 	  by some platforms.
 
+config BUILD_BOARD_IMAGE
```
maybe CUSTOM_IMAGE_BINARY

rgundi replied:
Done. Sounds better.

nashif commented on Kconfig.zephyr:
```diff
@@ -306,6 +306,13 @@ config BUILD_OUTPUT_STRIPPED
+config BUILD_BOARD_IMAGE
+	bool "Build a board-specific image"
+	help
+	  Build a board specific binary. This may differ from board to board
```
Document that a script is expected in the board directory under support

rgundi replied:
Done.

rgundi commented:
@nashif Addressed your comments.

sathishkuttan commented on boards/xtensa/intel_s1000_crb/support/create_board_img.py:
```diff
@@ -0,0 +1,221 @@
+
+To run script use command:
+
+create_spi_flash.py -i in_file.bin -o out_file.bin
```
The file name of the script is `create_board_img.py`

rgundi replied:
Thanks for catching.

sathishkuttan commented on boards/xtensa/intel_s1000_crb/support/create_board_img.py:
```diff
@@ -0,0 +1,221 @@
+FW_LOAD_NO_SHA_FLAG	= (1 << 25)
+FW_LOAD_CLK_SEL		= (1 << 21)
+
+FW_LOAD_ADDR		= (0xbe000000)
```
Can the load address be made configurable (through a command-line option, for example)? As we discussed, the bootloader's footprint is reduced to under 64KB, we would want to load the bootloader image on to LPSRAM.

rgundi replied:
sure. will see.

rgundi replied:
Done.

sathishkuttan reviewed (changes requested):
Thanks for the PR.
The script looks great.

rgundi commented on CMakeLists.txt:
```diff
@@ -1379,13 +1379,22 @@ list_append_ifdef(
   TARGET ${logical_target_for_zephyr_elf}
   POST_BUILD
   ${post_build_commands}
   COMMENT "Generating files from zephyr.elf for board: ${BOARD}"
```
Nope. This is unrelated to the change I am making.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -1379,13 +1379,22 @@ list_append_ifdef(
+  post_build_commands
+  COMMAND ${PYTHON_EXECUTABLE} ${BOARD_DIR}/support/create_board_img.py
+    -i ${PROJECT_BINARY_DIR}/zephyr.bin
+    -o ${PROJECT_BINARY_DIR}/zephyr_${BOARD}.bin
```
I am concerned that we will be polluting core build scripts, like zephyr/CMakeLists.txt with abstractions that are not flexible enough to be re-used many platforms.

e.g., what if a board needs to modify a hex file?
Which binary will a user expect to be flashed when he invokes "ninja flash" ?

Would you consider instead introducing this patch to zephyr/CMakeLists.txt:

```
get_property(
  extra_post_build_commands
  GLOBAL PROPERTY
  extra_post_build_commands
  )

list(APPEND
  post_build_commands
  ${extra_post_build_commands}
  )

```

Then you can inject the post_build_command from a board build script without needing to create a flexible abstraction.

rgundi replied:
Thanks for this suggestion, Sebastian. This is exactly what i wanted but didn't know how to do it. Will upload a revised patch.

rgundi committed:
CMakeLists: Introduce extra post build commands

This patchset introduces support for building board specific custom
images. Any board which needs to build a custom image (a post-processed
zephyr.bin) should inject the post_build_command from a script in the
corresponding board directory.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi commented:
@SebastianBoe @sathishkuttan Addressed your comments.

sathishkuttan commented on boards/xtensa/intel_s1000_crb/CMakeLists.txt:
```diff
@@ -2,3 +2,11 @@ if(CONFIG_PINMUX_INTEL_S1000)
+  COMMAND ${PYTHON_EXECUTABLE} ${BOARD_DIR}/support/create_board_img.py
+  -i ${PROJECT_BINARY_DIR}/zephyr.bin
+  -o ${PROJECT_BINARY_DIR}/zephyr_${BOARD}.bin
+  -l 0xbe000000
```
The load address is hardcoded, we need a way to override it from an application (specifically the bootloader app).

rgundi replied:
well.. this has to be same as "SRAM_BASE" defined in soc/xtensa/intel_s1000/memory.h coz the zephyr image generated for s1000 is not relocatable. So, somehow, i need to pass SRAM_BASE (or something similar) as an argument instead of hard-coding. Do you agree with this?

sathishkuttan replied:
Yes agreed. One option might to be take `zephyr.elf` as an input to the python script and parse the start address using `xt-readelf`.

rgundi replied:
Done.

sathishkuttan replied:
I see that you are leveraging pyelftools. pyelftools supports x86/x86_64 (and recently support ARM?), and does not explicitly support Xtensa. Because the usage is very simple (extract the start address from the header), it might be OK. Maybe we should just watch this, if you can add a console print of the address that is extracted from elf file, it would help detecting any wrong behavior with particular version(s) of tool/development host combination.

rgundi replied:
I added the print now anyway. Have a look.

rgundi commented:
@sathishkuttan Addressed your suggestion.
